core:driver implementation NXP CAAM Driver

cneveux opened the pull request:
Add the NXP CAAM drivers:
     - Random generator (instantiation and random generation)
     - Hash

   Add a generic cryptographic driver interface connecting
   TEE Crypto generic APIs to HW driver interface

cneveux commented:
@jenswi-linaro 

I submitted this patch to integrate the Crypto HASH done by NXP CAAM HW module. The current implementation of the Crypto operation is not fully filling our expectation. Could see the comment of the commit?

jenswi-linaro commented:
Hi @cneveux,

Thanks for this pull request. To better understand what's needed I'll need to ask a few questions and probably also propose different solutions.

> We would like to be able:
> - Select at runtime a SW algorithm or HW algorithm

Each time `crypto_hash_alloc_ctx()` is called or during boot after the driver has been initialized?
If each time `crypto_hash_alloc_ctx()` is called which criteria are used to tell if SW or HW is to be used, how will long used contexts be handled?

> - Select at compile time the best SW Library algorithm

That's already possible today for some definition of best. In our case it's preference of mbedtls or libtomcrypt which is the criterion. I assume that libtomcrypt generally gives better performance, but you could prefer mbedtls for other reasons. For instance company policy or plain trust.

cneveux commented:
> Hi @cneveux,
> 
> Thanks for this pull request. To better understand what's needed I'll need to ask a few questions and probably also propose different solutions.
> 
> > We would like to be able:
> > 
> > * Select at runtime a SW algorithm or HW algorithm
> 
> Each time `crypto_hash_alloc_ctx()` is called or during boot after the driver has been initialized?
> If each time `crypto_hash_alloc_ctx()` is called which criteria are used to tell if SW or HW is to be used, how will long used contexts be handled?

During HW Driver installation, the HW capacity is stored in the driver operation (max_hash) and function of this value, the Hash is computed by HW or SW. This is need in our case because CAAM Hash Limit is not the same function of the CAAM Module version.

We don't want to be only focus on the HASH, we will have the same runtime selection for other Crypto Alloc (e.g.  AES CCM/GCM/XTS/CTS, ...) Function of the CAAM version some mode are not supported.

Other point, when we will perform the crypto performance testing, we may need to add more criteria than the HW support. For example, we may prefer to do operation by SW because of the data size to be handled. 

Typically, when the amount of data is small, using HW is overloading the system with cache maintenance and buffer re-allocation to be cache line aligned (add memory copy too). In this case, doing the operation by SW is faster.
Cache alignment is need when HW DMA are used in write operation, the SW buffer in which we have to write must be cache aligned (start address and size of the buffer) to ensure that a cache invalidate is not affecting data around. If you refer to Linux, this is the role of the DMA memory allocator to do this. In OPTEE this feature is not supported as far as I know.

> 
> > * Select at compile time the best SW Library algorithm
> 
> That's already possible today for some definition of best. In our case it's preference of mbedtls or libtomcrypt which is the criterion. I assume that libtomcrypt generally gives better performance, but you could prefer mbedtls for other reasons. For instance company policy or plain trust.

Are we able to mix the Tomcrypt and mbedTLS? I mean, can I select the HASH384 with tomcrypt and HASH512 with mbedTLS? In fact the alloc_ctx should be upper to the tomcrypt or mbedTLS. May be having weak function.

In our vision, is to let end user (product manufacturer) to decide if he want to limit the runtime flexibility or not.

This could be benefict for every OPTEE customer.
(remember the libimxcrypt we shared in the past, it was done in this direction)

jenswi-linaro commented:
> During HW Driver installation, the HW capacity is stored in the driver operation (max_hash) and function of this value, the Hash is computed by HW or SW. This is need in our case because CAAM Hash Limit is not the same function of the CAAM Module version.

How about adding a call to `drvcrypt_hash_alloc_ctx()` at the beginning of `crypto_hash_alloc_ctx()` (in `core/crypto/crypto.c`) where we'll return directly if it succeeded and only fall back on the `switch (algo)` on failure?

This would mean that `crypto_{md5,sha1,sha224,sha256,sha384,sha512}_alloc_ctx()` are supposed to be the fallback software implementation, regardless of which crypto library is providing.

> We don't want to be only focus on the HASH, we will have the same runtime selection for other Crypto Alloc (e.g. AES CCM/GCM/XTS/CTS, ...) Function of the CAAM version some mode are not supported.

Let's find something we can agree on for hashes first. I agree that it could be an advantage to reuse the pattern, but I guess it also depends on the pattern.

> Other point, when we will perform the crypto performance testing, we may need to add more criteria than the HW support. For example, we may prefer to do operation by SW because of the data size to be handled.

> Typically, when the amount of data is small, using HW is overloading the system with cache maintenance and buffer re-allocation to be cache line aligned (add memory copy too). In this case, doing the operation by SW is faster.
Cache alignment is need when HW DMA are used in write operation, the SW buffer in which we have to write must be cache aligned (start address and size of the buffer) to ensure that a cache invalidate is not affecting data around. If you refer to Linux, this is the role of the DMA memory allocator to do this. In OPTEE this feature is not supported as far as I know.

The current API can't handle this. Is this something that can wait?

> Are we able to mix the Tomcrypt and mbedTLS? I mean, can I select the HASH384 with tomcrypt and HASH512 with mbedTLS? In fact the alloc_ctx should be upper to the tomcrypt or mbedTLS. May be having weak function.

Not mixing freely, but that could be achieved with more flags.

> In our vision, is to let end user (product manufacturer) to decide if he want to limit the runtime flexibility or not.

Sure, we'll accept patches. They should preferably be independent of this PR, I'm trying to limit the scope of this PR.

MrVan commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -48,6 +48,7 @@ else ifneq (,$(filter $(PLATFORM_FLAVOR),$(mx6ull-flavorlist)))
 $(call force,CFG_MX6,y)
 $(call force,CFG_MX6ULL,y)
 $(call force,CFG_TEE_CORE_NB_CORE,1)
+$(call force,CFG_NXP_CAAM,n)
```
"NXP" -> "IMX"? or this driver will also support LS chips?

sahilnxp replied:
Yes this driver will support LS chips also.

sdininno replied:
@MrVan ,
 yes the driver will support LS devices as well.

MrVan commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -65,13 +66,15 @@ else ifneq (,$(filter $(PLATFORM_FLAVOR),$(mx6s-flavorlist)))
 else ifneq (,$(filter $(PLATFORM_FLAVOR),$(mx7-flavorlist)))
 $(call force,CFG_MX7,y)
 CFG_TEE_CORE_NB_CORE ?= 2
+CFG_WITH_HAB ?= y
```
Please at least explain this macro some where, and why default set to y

MrVan commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -65,13 +66,15 @@ else ifneq (,$(filter $(PLATFORM_FLAVOR),$(mx6s-flavorlist)))
 $(call force,CFG_MX6,y)
 $(call force,CFG_MX6S,y)
 $(call force,CFG_TEE_CORE_NB_CORE,1)
+$(call force,CFG_NXP_CAAM,n)
```
Ditto

MrVan commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -184,6 +187,7 @@ $(call force,CFG_PL310,y)
 
 CFG_PL310_LOCKED ?= y
 CFG_ENABLE_SCTLR_RR ?= y
+CFG_WITH_HAB ?= y
```
Ditto

MrVan commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -244,3 +247,18 @@ CFG_SHMEM_SIZE ?= 0x00200000
+
+# Almost all platforms include CAAM HW Modules, except the
+# one force to be disabled
+CFG_NXP_CAAM ?= y
```
Will this break the image booting using vendor/upstream uboot + vendor/upstream linux? Or please share some public doc somewhere to use the image built out with this "y"

cneveux replied:
removed

MrVan commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -244,3 +247,18 @@ CFG_SHMEM_SIZE ?= 0x00200000
+CFG_NXP_CAAM ?= y
+
+ifeq ($(CFG_NXP_CAAM),y)
+# currently disable the use of CAAM in OP-TEE
```
Please add a comment about why "currently disable ..."

cneveux replied:
added explanation

MrVan commented on core/arch/arm/plat-imx/crypto_conf.mk:
```diff
@@ -0,0 +1,52 @@
+#
+ifeq ($(CFG_CRYPTO_DRIVER), y)
+#
+# Define this variable ifthe system is able to manage the
```
Typo "ifthe"

MrVan commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,156 @@
+#include "hal_cfg.h"
+#include "hal_clk.h"
+#include "hal_ctrl.h"
+
```
Please sort the header

MrVan commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,156 @@
+#include "hal_clk.h"
+#include "hal_ctrl.h"
+
+#define CTRL_DEBUG
```
Prefer a CFG option

cneveux replied:
rework debug trace in the driver to select the traces function of build CFG_CAAM_DBG 32 bits word

Emantor commented:
How do you propose to solve the problem of the linux clock driver turning of the clocks to the CAAM if it is used within OP-TEE but not enabled in the Linux kernel?
The Linux kernel will turn of the CAAM clocks even if they are enabled during the CAAM initialization, since there is no consumer visible to the clock framework.

MrVan commented:
@Emantor If caam clk disabled by Linux, accessing it in OP-TEE, OP-TEE will crash, secure data will not be leaked. So does clk really matters here?

Emantor commented:
Yes, you won't leak data, but your OP-TEE and thus your device is not usable as well.
Since the CAAM clks are disabled your transaction is stuck on the bus and your device is hard locked until the watchdog stikes.

MrVan commented:
> Yes, you won't leak data, but your OP-TEE and thus your device is not usable as well.
> Since the CAAM clks are disabled your transaction is stuck on the bus and your device is hard locked until the watchdog stikes.

There is no good method to avoid CCM module being hacked in Linux side, unless protect CCM in secure world, but this requires redesign Linux side clock driver...
Not sure whether other silicon vendors has such issue and how.

sdininno commented:
> How do you propose to solve the problem of the linux clock driver turning of the clocks to the CAAM if it is used within OP-TEE but not enabled in the Linux kernel?
> The Linux kernel will turn of the CAAM clocks even if they are enabled during the CAAM initialization done by OP-TEE, since there is no consumer visible to the clock framework.

Currently, as @MrVan stated, the clk will not be managed by Linux. Off course as you have stated rightfuly it is subject to a DoS attack. Unfortunatly , the TEE is not design fot that. it  rather guarantees data integrity, not cpu availability.
having said that, our plan is also to move clock control to TEE (at least for i.MX6Q/DL, but could expand to all the 6 and 7 as well) and to expose an interface to Linux. we do have patches for that but are not ready yet for integration.

sdininno commented:
Hello @jenswi-linaro , 
> 
> How about adding a call to `drvcrypt_hash_alloc_ctx()` at the beginning of `crypto_hash_alloc_ctx()` (in `core/crypto/crypto.c`) where we'll return directly if it succeeded and only fall back on the `switch (algo)` on failure?
> 
This is good alternative, we will look at this option.
just to let you know that @cneveux  will not be working on this for some time, but i will try to continue this work.

Emantor commented on core/drivers/crypto/caam/hal/common/hal_cfg.c:
```diff
@@ -0,0 +1,293 @@
+
+#ifdef CFG_DT_MODIFY
+	/* Ensure that CAAM Control is secure-status enabled */
+	if (dt_set_secure_status(fdt, node)) {
```
`dt_set_secure_status` does not exist in this patch set nor in the OP-TEE master.

cneveux replied:
Correct. This is why it's built only if the CFG_DT_MODIFY is define. This will  require a specific patch to add this function. Do you agree to keep it for now while a new patch is submitterd?

Emantor replied:
Either remove this function or add the implementation, this is up to you. However I think most of the dt modification functions contained in the HAL are general enough to be included in the OP-TEE core files instead of the CAAM hal.

cneveux commented:
Hello @jenswi-linaro 

I reworked the patch to add the function drvcrypt_hash_alloc_ctx() at the beginning of crypto_hash_alloc_ctx().
I removed the crypto SW wrapper from the original PR.

dmcilvaney commented:
Hi @cneveux , @MrVan,

Is an implementation of `tee_otp_get_hw_unique_key()` and `tee_otp_get_die_id()` based on the OTPMK outside the scope of this pull request, or something you would consider including? We would love to get your implmementation [from Code Aurora](https://source.codeaurora.org/external/imx/imx-optee-os/tree/core/arch/arm/plat-imx/imx_huk.c?h=rel_imx_4.14.98_2.0.0_ga&id=7edb46e0fe60261d06bb2390b716201d6772e16f)

We have an implementation we have been using [here](https://github.com/ms-iot/optee_os/blob/ms-iot/core/drivers/fsl_sec/hw_key_blob.c) which uses the CAAM to generate them, but ideally we would like to move completely to an upstream implementation rather than maintaining our partial CAAM driver.

cneveux commented:
> Hi @cneveux , @MrVan,
> 
> Is an implementation of `tee_otp_get_hw_unique_key()` and `tee_otp_get_die_id()` based on the OTPMK outside the scope of this pull request, or something you would consider including? We would love to get your implmementation [from Code Aurora](https://source.codeaurora.org/external/imx/imx-optee-os/tree/core/arch/arm/plat-imx/imx_huk.c?h=rel_imx_4.14.98_2.0.0_ga&id=7edb46e0fe60261d06bb2390b716201d6772e16f)
> 
> We have an implementation we have been using [here](https://github.com/ms-iot/optee_os/blob/ms-iot/core/drivers/fsl_sec/hw_key_blob.c) which uses the CAAM to generate them, but ideally we would like to move completely to an upstream implementation rather than maintaining our partial CAAM driver.

@dmcilvaney,
This is not in the scope of this PR. We would like first to add all NXP CAAM support. Next we will have other PRs to add such features. So the HUK support will be in separate PR

jenswi-linaro commented:
For the commit "core:arm64 Add sev/wfe inline function":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on MAINTAINERS:
```diff
@@ -102,6 +102,13 @@ R:	Peng Fan <peng.fan@nxp.com> [@MrVan]
 F:	core/arch/arm/plat-imx/
 
+NXP (Freescale) i.MX family
+R:	Cedric Neveux <cedric.neveux@nxp.com>
```
Consider adding Github id as is done for many other platforms.

cneveux replied:
ok, I'll do it

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,101 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+	 */
+	int_prop = fdt_getprop(fdt, node, "interrupts", &len_prop);
+
+	if ((int_prop == NULL) || (len_prop < 2))
```
It's a matter of taste, but I think that `(int_prop == NULL)` looks a bit clumsy compared to just `!int_prop`. Do as you like.

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,101 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+}
+
+paddr_t dt_node_offset_by_compatible_status(void *fdt, int status, int *node,
+		const char *compatible)
```
Please align after the `(` at the line above.

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,101 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
 
+int dt_get_irq(void *fdt, int node)
+{
+	const uint32_t *int_prop;
```
Please initialize all local (automatic) variables in this function and in the other new functions.

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,101 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+	int node_status;
+
+	fnode = fdt_node_offset_by_compatible(fdt, *node,
+			compatible);
```
Please align after the `(` at the line above, here and at other places in this file.

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,101 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+		if (fdt_setprop_string(fdt, node, "status", "disabled"))
+			return -1;
+	} else {
+		/* status is there, modify it */
```
```
/*
 * Please format
 * multi-line
 * comments
 * like this.
 */

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,101 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+		 * not "ok" or "okay" will work
+		 */
+		if (fdt_setprop_inplace(fdt, node, "status",
+				"disa", len))
```
I suppose you're using a truncated status value as an optimisation to avoid growing the FDT, you can mention that in the comment.
What if `len` is larger than `5`?

cneveux replied:
I'll update the comment with more explanation. 
The goal is to modify the DT with a value not equal to "ok" or "okay". I'll modify the new setting value to be "disabled" because the fdt_setprop_inplace is going to truncate the value function of the "len" value and else copy the "disabled" at the beginning of the property and keep the rest of the original property.
Objectif of using this function is to not change the DT offset and so recalculate the overall DT. For me, this is a time consuming operation that is not necessary in this case.

jenswi-linaro commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,101 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+	} else {
+		/* status is there, modify it to okay */
+		if (fdt_setprop_inplace(fdt, node, "secure-status",
+				"okay", len))
```
What if `len` is larger than `5`?
`secure-status` must be set to `ok` or `okay`, what if `len` is `< 2` or `3`?

cneveux replied:
I'll change the function to use the `fdt_appendprop_string` that is taking care of either creating the entry or modifying it by setting the "okay" value and updating property length too.

cneveux commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,130 @@
+#endif
+
+/* Local includes */
+#include "caam_common.h"
```
@jenswi-linaro 
Agree with you in other implementation when the include files are in global include directory, but in this case, the `core/drivers/crypto/caam/include` are kept locally to the `core/driver/crypto/caam `sources. 

I didn't want to make those include accessible from other sources as they are used only in the caam driver implementation. The path is not global-incdir in the sub.mk, this is why I kept the local include definition. There is nothing exported and to be re-used by other part of code elsewhere. Files are completely private to the local driver implementation.

Now if you prefer me to moved them in the `core/include/drivers/crypto/caam`, I'll do it and so prefix all include file name with `caam_`. 

Please let me know if you still want me to move file in global include directory.

jenswi-linaro replied:
I understand, no need to move the .h files to a global include directory.

However, the files included via the include paths (using `-I` argument to the compiler) should still preferably be included with `#include <file.h>` rather than `#include "file.h"`. All files included via include paths can cause conflicts so prefixing is still needed, unless they are included as `#include <caam/file.h>`, then `caam/` is the prefix.

cneveux replied:
ok understood. I'll add file prefix and change the `#include"file.h"` to `#include <file.h>`

cneveux commented on core/crypto/crypto.c:
```diff
@@ -14,30 +14,41 @@
+	 * the driver's allocation function. If return no success
+	 * call the SW defined library
+	 */
+	res = drvcrypt_hash_alloc_ctx(&c, algo);
```
I'll do

cneveux commented:
> > Hi @cneveux , @MrVan,
> > Is an implementation of `tee_otp_get_hw_unique_key()` and `tee_otp_get_die_id()` based on the OTPMK outside the scope of this pull request, or something you would consider including? We would love to get your implmementation [from Code Aurora](https://source.codeaurora.org/external/imx/imx-optee-os/tree/core/arch/arm/plat-imx/imx_huk.c?h=rel_imx_4.14.98_2.0.0_ga&id=7edb46e0fe60261d06bb2390b716201d6772e16f)
> > We have an implementation we have been using [here](https://github.com/ms-iot/optee_os/blob/ms-iot/core/drivers/fsl_sec/hw_key_blob.c) which uses the CAAM to generate them, but ideally we would like to move completely to an upstream implementation rather than maintaining our partial CAAM driver.
> 
> @dmcilvaney,
> This is not in the scope of this PR. We would like first to add all NXP CAAM support. Next we will have other PRs to add such features. So the HUK support will be in separate PR

@dmcilvaney 
I added a preliminary patch implementing the CAAM driver to generate a HUK in the PR #3149

jenswi-linaro commented:
For the commit "core:dt add DT function to read/modify DT" please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

Emantor commented:
> > How do you propose to solve the problem of the linux clock driver turning of the clocks to the CAAM if it is used within OP-TEE but not enabled in the Linux kernel?
> > The Linux kernel will turn of the CAAM clocks even if they are enabled during the CAAM initialization done by OP-TEE, since there is no consumer visible to the clock framework.
> 
> Currently, as @MrVan stated, the clk will not be managed by Linux. Off course as you have stated rightfuly it is subject to a DoS attack. Unfortunatly , the TEE is not design fot that. it rather guarantees data integrity, not cpu availability.
> having said that, our plan is also to move clock control to TEE (at least for i.MX6Q/DL, but could expand to all the 6 and 7 as well) and to expose an interface to Linux. we do have patches for that but are not ready yet for integration.

Currently, these patches are incompatible with the mainline Linux kernel. The kernel clock driver will automatically turn off the CAAM clocks if the CAAM driver is not enabled within the kernel. Moving clock control to TEE is a solution to the problem, but requires coordination with the upstream kernel maintainers on how these interfaces are supposed to be designed. Do you have links to the discussion with upstream for this?

cneveux commented:
> > > How do you propose to solve the problem of the linux clock driver turning of the clocks to the CAAM if it is used within OP-TEE but not enabled in the Linux kernel?
> > > The Linux kernel will turn of the CAAM clocks even if they are enabled during the CAAM initialization done by OP-TEE, since there is no consumer visible to the clock framework.
> > 
> > 
> > Currently, as @MrVan stated, the clk will not be managed by Linux. Off course as you have stated rightfuly it is subject to a DoS attack. Unfortunatly , the TEE is not design fot that. it rather guarantees data integrity, not cpu availability.
> > having said that, our plan is also to move clock control to TEE (at least for i.MX6Q/DL, but could expand to all the 6 and 7 as well) and to expose an interface to Linux. we do have patches for that but are not ready yet for integration.
> 
> Currently, these patches are incompatible with the mainline Linux kernel. The kernel clock driver will automatically turn off the CAAM clocks if the CAAM driver is not enabled within the kernel. Moving clock control to TEE is a solution to the problem, but requires coordination with the upstream kernel maintainers on how these interfaces are supposed to be designed. Do you have links to the discussion with upstream for this?

@Emantor 
We agree on this point. Today the Linux is enabling the CAAM driver hence we have no difficulty to enable the CAAM driver in the TEE as done with this PR. A Linux and associated TEE patches have been implemented and are internally under review. Unfortunately it takes time to get approval for such big changes. We have to live with what we have available today.

jenswi-linaro commented on core/drivers/crypto/caam/utils/utils_mem.c:
```diff
@@ -0,0 +1,539 @@
+/**
+ * @brief   Align the pointer \a p to be align with \a align modulus
+ */
+#define PTR_ALIGN(p, align)	(p + ((align - (p % align)) % align))
```
I don't understand what is macro does. Is it supposed to round up or down or something else?

jenswi-linaro commented on core/drivers/crypto/caam/utils/utils_mem.c:
```diff
@@ -0,0 +1,539 @@
+/**
+ * @brief   Cache line size in bytes
+ */
+static uint16_t cacheline_size;
```
I don't think we need a global variable for this. Reading a system register is cheap enough that we read it when needed.

jenswi-linaro commented on core/drivers/crypto/caam/utils/utils_mem.c:
```diff
@@ -0,0 +1,539 @@
+ * @brief   Read the system cache line size.\n
+ *          Get the value from the ARM system configration register
+ */
+static void read_cacheline_size(void)
```
This should rather be an inline function in arm32.h and arm64.h.

jenswi-linaro commented on core/drivers/crypto/caam/utils/utils_mem.c:
```diff
@@ -0,0 +1,539 @@
+	uint32_t value;
+
+#ifdef ARM64
+	asm volatile ("mrs %0, ctr_el0" : "=r" (value));
```
Define a `read_ctr_el0()` in arm64.h and use that instead.

jenswi-linaro commented on core/drivers/crypto/caam/utils/utils_mem.c:
```diff
@@ -0,0 +1,539 @@
+	cacheline_size = 4 << ((value >> CTR_DMINLINE_SHIFT)
+						& CTR_DMINLINE_MASK);
+#else
+	asm volatile ("mrc p15, 1, %0, c0, c0, 0" : "=r" (value));
```
This assumes that the correct cache has been selected already. Why not use `read_ctr()` instead?

cneveux commented:
@jenswi-linaro, @etienne-lms 

I reviewed the PR according your comments in the different PRs I submitted. I changed the overall comments to remove Doxygen and I added the initialization of  all variables (hoping I didn't miss one)
I reworked the DT functions according to @etienne-lms review (I hope this is corresponding to your expectation).

etienne-lms commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,99 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+	 * If the property is not present, it will be added
+	 * otherwise it will be modified to be "okay"
+	 */
+	if (fdt_setprop_string(fdt, node, "secure-status", "okay"))
```
what if secure-status already exists?
This should do the same trick as done in `dt_disable_status()`. Even a bit more. Case string is "d\0", there's no room to change in place into "ok\0".

cneveux replied:
if secure-status already exists the value will be forced to `okay`. We can not apply the same as dt_disable_status because to disable a status, writing something different of `ok` or `okay` is doing the job. Here we want to enable the status so we have to set it explicitly to `ok` or `okay`. Choice has been done to set it to `okay`. (refer to `is_okay()` in the same file.

etienne-lms replied:
Ok, fine for setting `secure-status` to `okay`. 
That said, you still face with `secure-status` the same problematic as faced with `status`. If `secure-status` is already defined but shorter than 6 chars (including `\0`) you must take care not to override DT data by loading string "okay".

cneveux replied:
For me this is not obvious. The `fdt_setprop_string` is not supposed to do the job for us? In the `fdt_setprop_string` description it's said "create or change a property" and "This function may insert or delete data from the blob, and will therefore change the offsets of some existing nodes"
So what do you want me to do more. Do you want me the check if the `secure-status` already exist, then I have to delete it and recreate it?

etienne-lms replied:
If `fdt_setprop_string(fdt, node, "secure-status", "okay")` is fine to force "secure-status" to "okay", then you should be able to implement `dt_disable_status()` the same way:
```C
int dt_disable_status(void *fdt, int node)
{
	if (fdt_setprop_string(fdt, node, "status", "disabled"))
		return -1;

 	return 0;
}
```

cneveux replied:
Yes, but do we want to spend time in recalculating all the Device Tree when the `status` property is already present and just changing the `ok` or `okay` with `di` or `disab` is disabling it without recalculating all the device tree?
Let me know.

etienne-lms replied:
I get your point. I still think it's a bit strange to see those 2 functions with such different implementation whereas they have about the same purpose.

Minor: can remove the inline comment in `dt_enable_secure_status()`, it adds no info.

cneveux commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -251,3 +252,21 @@ CFG_SHMEM_SIZE ?= 0x00200000
+# it with generic crypto API can be enabled.
+CFG_CRYPTO_DRIVER ?= y
+# Crypto Driver Debug
+CFG_CRYPTO_DRV_DBG ?= n
```
ok

cneveux commented on core/drivers/crypto/crypto_api/include/drvcrypt_hash.h:
```diff
@@ -0,0 +1,54 @@
+	HASH_SHA256,        ///< SHA 256
+	HASH_SHA384,        ///< SHA 384
+	HASH_SHA512,        ///< SHA 512
+	MAX_HASH_SUPPORTED  ///< Maximum Hash supported
```
ok

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,290 @@
+	return ret;
+}
+
+const struct crypto_hash_ops hash_ops = {
```
ok

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,290 @@
+const struct crypto_hash_ops hash_ops;
+
+/**
+ * @brief   Checks if input API context is correct. If not, system break.
```
ok

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,290 @@
+	uint8_t algo_id;
+	enum drvcrypt_hash_id hash_id;
+
+	/* Extract the algorithms fields */
```
ok

cneveux commented on core/drivers/crypto/crypto_api/include/drvcrypt_hash.h:
```diff
@@ -0,0 +1,54 @@
+ */
+struct drvcrypt_hash {
+	///< Maximum hash supported
+	enum drvcrypt_hash_id max_hash;
```
I'll change the variable type but I'll not remove it. Some HW are not able to manage all HASH sizes. Hence I need to keep this information

cneveux replied:
Sorry I see your point, I'll remove it and let the HW driver to check it case by case

etienne-lms replied:
fine, thanks.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+ * Enable the interrupt mode or not
+ */
+#ifdef CFG_CRYPTO_DRIVER
+//#define IT_MODE
```
Some leftover?

cneveux commented on core/drivers/crypto/crypto_api/include/drvcrypt_hash.h:
```diff
@@ -0,0 +1,54 @@
+	TEE_Result (*final)(void *ctx, uint8_t *digest, size_t len);
+
+	///< Copy Hash context
+	void (*cpy_state)(void *dst_ctx, void *src_ctx);
```
ok

cneveux commented on core/drivers/crypto/crypto_api/include/drvcrypt_hash.h:
```diff
@@ -0,0 +1,54 @@
+	void (*cpy_state)(void *dst_ctx, void *src_ctx);
+
+	///< HMAC Key computing
+	TEE_Result (*compute_key)(void *ctx, const uint8_t *key, size_t len);
```
true, I'll remove it

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,290 @@
+		free(hash);
+		*ctx = NULL;
+	} else {
+		hash->hash_ctx.ops = &hash_ops;
```
This is how the crypto_hash_* is working except if I misunderstood something

etienne-lms replied:
Ok, sorry, I got confused between the HW hash driver, the drvcrypto_hash drivers and the crypto_hash layer.

cneveux commented on core/include/kernel/dt.h:
```diff
@@ -93,6 +95,51 @@ int dt_map_dev(const void *fdt, int offs, vaddr_t *base, size_t *size);
+ *
+ * @fdt reference to the Device Tree
+ * @status is the status flag of the node to found
+ * @node is the node offset to start from in DTB
```
Sure

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+/*
+ * Definition of output ring object
+ */
+struct __attribute__((__packed__)) outring_entry {
```
Please use the `__packed` macro from `<compiler.h>` instead.

cneveux replied:
ok

cneveux commented on core/include/kernel/dt.h:
```diff
@@ -93,6 +95,51 @@ int dt_map_dev(const void *fdt, int offs, vaddr_t *base, size_t *size);
+ * If found check if its status value fits the given @status value.
+ *
+ * @fdt reference to the Device Tree
+ * @status is the status flag of the node to found
```
ok

cneveux commented on core/include/kernel/dt.h:
```diff
@@ -93,6 +95,51 @@ int dt_map_dev(const void *fdt, int offs, vaddr_t *base, size_t *size);
+		const char *compat);
+
+/*
+ * Modify or add "status" field of property to "disabled"
```
ok

cneveux commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,99 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+
+	while (fnode != -FDT_ERR_NOTFOUND) {
+		node_status = _fdt_get_status(fdt, fnode);
+		if (node_status & status) {
```
I agree to remove the node_status but I fully disagree on the equality check.
The `_fdt_get_status` can return (DT_STATUS_OK_NSEC | DT_STATUS_OK_SEC) and we want to check one flag or the other.

cneveux commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,99 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+
+	prop = fdt_getprop(fdt, node, "status", &len);
+	if (!prop) {
+		/* Status is not available, just add it */
```
ok

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+ * Definition of output ring object
+ */
+struct __attribute__((__packed__)) outring_entry {
+	descEntry_t  desc;   /* Physical address of the descriptor */
```
So this struct becomes either
```
struct __packed {
    uint64_t desc;
    uint32_t status;
};
```
or
```
struct __packed {
    uint32_t desc;
    uint32_t status;
};
```

Depending on `CFG_CAAM_64BITS`. The conditional typedeffing in another file even hides this quite well and it's not obvious if it's intended behaviour or not either.

cneveux replied:
This is expected. If CAAM HW is handling 64 bits addressing mode the CFG_CAAM_64BITS is defined and all physical addresses handled by the CAAM HW must be 64 bits and not 32 bits. There is no link between the ARM architecture and the CAAM Addressing mode.

cneveux replied:
Yes, you're correct.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+struct caller_info {
+	struct jr_jobctx *jobctx; /* Caller job context object */
+	uint32_t         jobid;   /* Current Job Id */
+	descEntry_t      pdesc;   /* Physical address of the descriptor */
```
We're normally using `paddr_t` for physical addresses.

cneveux replied:
because this is related to the CAAM addressing mode in this case

cneveux replied:
ok

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+
+	vaddr_t ctrladdr;               /* CAAM virtual base address */
+	paddr_t jroffset;               /* Job Ring address offset */
+	uint64_t paddr_inrings;         /* Physical address of input queue */
```
Why aren't these `paddr_t` instead?

cneveux replied:
This is the HW CAAM configuration registers that are 64 bits coded.

cneveux replied:
Yes to get the physical address but HW is 2x32 bits registers for this. I'll add comments.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+				/* Clear the Entry Descriptor DMA */
+				caller->pdesc = 0;
+				caller->jobid = JR_JOB_FREE;
+				found = idx_jr;
```
Why is this needed when `jobctx` is assigned as well?
Putting this for loop in a helper function which would return the found `jobctx` would simplify this function a bit.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+	struct caller_info *caller = NULL;
+	uint32_t exceptions = 0;
+	uint32_t job_mask = 0;
+	uint8_t  idx_jr = 0;
```
I'd prefer `uint8_t idx_jr = 0;` without the extra space, same on the row below.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+	 * also touching it
+	 */
+	cpu_spin_lock(&jr_privdata->callers_lock);
+	for ((idx_jr = 0), (found = false);
```
With this loop in a helper function returning the found `struct caller_info *` would simplify things a bit.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+		if (jr_privdata->callers[idx_jr].jobid == JR_JOB_FREE) {
+			JR_TRACE("Found a space #%d free in the callers array",
+				idx_jr);
+			job_mask = 1 << idx_jr;
```
Prefer using the `BIT32()` macro from `<util.h>`

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+
+	/* Initialize the spin locks */
+	jr_priv->inlock = SPINLOCK_UNLOCK;
+	jr_priv->outlock = SPINLOCK_UNLOCK;
```
Unused?

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+		 * Lock the caller information array because enqueue is
+		 * also touching it
+		 */
+		cpu_spin_lock(&jr_privdata->callers_lock);
```
It seems that this lock i only taken when already holding `inlock`. So this is perhaps not needed, or will it be needed later on?

cneveux replied:
This spin lock is to protect the access to the jr_privdata->callers array that can be modified in same time when a new Job is enqueued.
This is necessary (and tested) in case of mutli-core. One core can push a new job while another fetches the completed jobs.

jenswi-linaro replied:
In the current state of the code you're always holding `inlock` before you're taking `callers_lock`. So I'd say that it's not needed right now, unless there's a bug in the code. If it will be needed later on as the code will evolve with more support it's fine to keep it.

cneveux replied:
Ah yes, correct my mistake I'll change to use the outlock spinlock object in the dequeue.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+	/* Ensure that physical memory is up to date */
+	cache_operation(
+		TEE_CACHECLEAN,
+		(void *)(&jr_privdata->inrings[jr_privdata->inwrite_index]),
```
This is a bit much, can we use a helper variable or something?

cneveux replied:
ok

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+
+	/* Push the descriptor into the JR HW list */
+	caam_write_val(&(jr_privdata->inrings[jr_privdata->inwrite_index]),
+		caller->pdesc);
```
Please align after the `(` above.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+	}
+
+	JR_TRACE("Push id=%d, job (0x%08x) context @0x%08"PRIxVA"",
+			jr_privdata->inwrite_index, job_mask, (vaddr_t)jobctx);
```
Please align after the `(` above. I think there's some option for checkpatch to complain over this, I'm not sure if it's `-strict` or something else.

cneveux replied:
Problem is when you use tab with 4 or 8 spaces you don't have the same result.
Checkpatch is tab=8 spaces and I'm editing file with tab=4 spaces (because I prefer).
I don't know if there is a checkpatch option for that but for sure checkpatch is complaining if a line cross the 80 characters.

etienne-lms replied:
optee_os uses the Linux kernel coding style: https://optee.readthedocs.io/general/coding_standards.html.
I should update your file editor to use 8-char tabs when in optee_os source tree, as you probably do when browsing the Linux kernel source tree.

jenswi-linaro commented on core/drivers/crypto/caam/include/caam_status.h:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Internal CAAM Driver status codes
+ */
+enum CAAM_Status {
```
I'm wondering if this is needed or we could use `TEE_Result` instead.
If we're translating any of these values into `TEE_Result` values we could save a bit. There are other drivers that uses `TEE_Result` instead of a enum defined for just that particular driver.

cneveux replied:
This is something we introduce following internal NXP security expert groups. This code is implemented since a long time and I'm reworking it for the nth time. If you really want me to remove that I can do. Please advise.

jenswi-linaro replied:
This isn't very important. Right now you don't seem to translate this into different TEE_Result values very often. If you prefer keeping as it is, do so.

etienne-lms commented on core/drivers/crypto/crypto_api/hash/sub.mk:
```diff
@@ -0,0 +1,2 @@
+
```
minor: can remove this empty line

cneveux replied:
ok

etienne-lms commented on core/drivers/crypto/crypto_api/include/drvcrypt.h:
```diff
@@ -0,0 +1,69 @@
+#ifndef __DRVCRYPT_H__
+#define __DRVCRYPT_H__
+
+/* Global includes */
```
remove this comment

cneveux replied:
ok

etienne-lms commented on core/drivers/crypto/crypto_api/include/drvcrypt.h:
```diff
@@ -0,0 +1,69 @@
+ * 0   Success
+ * (-1) Error
+ */
+int drvcrypt_register(enum drvcrypt_algo_id idx, void *ops);
```
This should rather return a `TEE_Result` value.

cneveux replied:
ok

etienne-lms commented on core/drivers/crypto/crypto_api/include/drvcrypt.h:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Cryptographic module registration
+ *
+ * Inputs:
```
prefer without `Inputs:` and with `@` to highlight argument labels and some rephrasing (s/idx/algo_id/).
```
/*
 * Cryptographic module registration
 *
 * @id  Identifier for the cryptographic algorithm family 
 * @ops  Registered cryptographic module operation handlers
 ```

cneveux replied:
ok I'll review everywhere

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,310 @@
+ * Outputs:
+ * digest  Hash digest buffer
+ *
+ * Returns:
```
minor: no real need to detail the return values here.

cneveux replied:
Could you please say me what are the function header rules you apply?

etienne-lms replied:
As you can see browsing optee_os source files there is no unique way for function description.

cneveux commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,781 @@
+	/* Check if there is an available JR index in the HW */
+	while (hal_jr_read_nbSlotAvailable(jr_privdata->baseaddr) == 0) {
+		/*
+		 * WFE will be exit be a SEV generated but the
```
ok

jenswi-linaro commented on core/include/io.h:
```diff
@@ -94,6 +94,26 @@ static inline void put_be16(void *p, uint16_t val)
+
+static inline uint32_t get_le32(const void *p)
+{
+	return (*(const uint32_t *)p);
```
The outer `(` and `)` are not needed

cneveux replied:
ok

jenswi-linaro commented on core/include/io.h:
```diff
@@ -94,6 +94,26 @@ static inline void put_be16(void *p, uint16_t val)
+
+static inline uint64_t get_le64(const void *p)
+{
+	return (*(const uint64_t *)p);
```
The outer `(` and `)` are not needed

cneveux replied:
ok

etienne-lms commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,749 @@
+	jr_priv->callers_lock = SPINLOCK_UNLOCK;
+
+	/* Initialize the queue counter */
+	jr_priv->inwrite_index = jr_priv->outread_index = 0;
```
prefer split over 2 lines:
```
	jr_priv->inwrite_index = 0;
	jr_priv->outread_index = 0;
```
Note that this is not required since `caam_ammoc*()` zeroes allocated memory.

cneveux replied:
ok to split, but I prefer to keep initialization as the caam_alloc is function of a define that could be changed.

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,253 @@
+	struct crypto_hash *hash_src = to_hash_ctx(src_ctx);
+	struct crypto_hash *hash_dst = to_hash_ctx(dst_ctx);
+
+	if ((!dst_ctx) || (!src_ctx))
```
should be check before `to_hash_ctx()` is called.

cneveux replied:
In fact you pointed an error, this is the hash_src and hash_dst that must be checked. The to_hash_ctx is already doing the check of dst_ctx and src_ctx.

etienne-lms replied:
Actually I think `hash_src` as well as `hash_dst` cannot be `NULL`. Maybe `dst_ctx` (and/or `src_ctx`) can be null.
This applies also to `hash_free_ctx()` above and `hash_{init|update|final}()` below.

Note: I one calls `hash_copy_state()` when  `hash_src->op->copy_state == NULL`, this looks like a bug and it should be reposed somehow.

cneveux replied:
Ok, so what do you prefer? 

1. Keep the `to_hash_ctx` as is with the `assert` and remove the check of the `to_hash_ctx` returned?
2. Remove the `assert` and checked if pointer are not NULL?

Then I would say that a driver might not have to implement all operations. The copy_state is fully related to the driver context hence if we don't need to manage a context there is no need for a copy. Again, if you prefer to have all driver functions implemented even if they are empty, I can change this file to panic if a function pointer is NULL

etienne-lms replied:
I would say option 1.

As for the copy-state function, i think you should do as for [`crypto_hash_copy_state()`](https://github.com/OP-TEE/optee_os/blob/3.6.0/core/crypto/crypto.c#L64).

cneveux replied:
ok

cneveux commented on core/kernel/dt.c:
```diff
@@ -51,6 +51,93 @@ bool dt_have_prop(const void *fdt, int offs, const char *propname)
+	fnode = fdt_node_offset_by_compatible(fdt, node, compat);
+
+	while (fnode != -FDT_ERR_NOTFOUND) {
+		if (_fdt_get_status(fdt, fnode) & status)
```
@etienne-lms 
I'm sorry but I don't understand. 
The problem is not the `status` but the `_fdt_get_status` that returns `DT_STATUS_NSEC | DT_STATUS_SEC` if `status = okay` and `secure-status = okay`.
As properties are defined in the DTB, we can not control that. What the function is doing is to check if `_fdt_get_status` returned value as a `status` bit set or not.

etienne-lms replied:
I think one could legitimately call this function with status=`DT_STATUS_DISABLED` to find a compatible node where both status and secure-status are disabled. Your implementation suggests that this is not a valid scenario.

Do you expect this function to find a node with exclusively one of `status` and `secure-status` being enabled, the other being disable, and never with both enable?

I still think a simpler, explicit and generic way is to expect target node get `_fdt_get_status()` to exactly match input argument `status`.

Or if there is no clean generic way, maybe move this function as a local function of your driver that needs it.

cneveux replied:
You're right on the fact that the function is not handling the DT_STATUS_DISABLED. This is no so complex to add this case (I just didn't think about it).

Next for me the biggest problem is coming from the `_fdt_get_status` function itself in which I don't see why the `DT_STATUS_SEC` is returned because `status=okay` and `secure-status` is not present. This is a blocking point because their is no possibility to distinguish if the node is only secure.
In same cases, we can not considered that a node is secure because `status=okay`, some node should be exclusively Secure or Non-Secure.

See that we are discussing since a long time on this function and if you agree I'll move it back to my specific driver as it was before. Do you agree?

etienne-lms replied:
> Next for me the biggest problem is coming from the _fdt_get_status function itself in which I don't see why the DT_STATUS_SEC is returned because status=okay and secure-status is not present. This is a blocking point because their is no possibility to distinguish if the node is only secure.

This behavior follows the DT bindings. No `status` is also equivalent to `status = "okay"`. A node enabled only for secure world should explicitly state:
```
    status = "disabled"; 
    secure-status = "okay";
```

cneveux replied:
remove this function from generic code

cneveux commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,119 @@
+		hal_cfg_setup_nsjobring(jr_cfg.base);
+
+	CTRL_TRACE("CAAM Driver initialization (0x%x)\n", retresult);
+	return retresult;
```
fine, I'll do

cneveux commented on core/drivers/crypto/caam/caam_desc.c:
```diff
@@ -0,0 +1,90 @@
+	caam_write_val((void *)last, word);
+
+	/* Increase the length */
+	caam_write_val((void *)(desc), (caam_read_val((void *)desc) + 1));
```
yes

cneveux commented on core/drivers/crypto/caam/include/caam_desc_defines.h:
```diff
@@ -0,0 +1,658 @@
+/*
+ * Common Command constants
+ */
+#define CMD_TYPE(cmd)			SHIFT_U32((cmd & 0x1F), 27)
```
ok

cneveux commented on core/drivers/crypto/caam/include/caam_desc_defines.h:
```diff
@@ -0,0 +1,658 @@
+	ECC_P384,
+	ECC_P521,
+	ECC_MAX,
+	ECC_UNKNOWN = (0xFF),
```
I'll change the enumerate but as you said this only apply to the CAAM and file is not globally exported. I prefer to keep rest as is because it will increase macro name size that are not to be used outside the CAAM driver.

cneveux commented on core/drivers/crypto/caam/include/caam_desc_helper.h:
```diff
@@ -0,0 +1,516 @@
+/*
+ * Descriptor pointer type
+ */
+typedef uint32_t *descPointer_t;
```
This is no exactly that. 
The CAAM is working with physical addresses that can be either 32 or 64 bits.
Then each entry of the descriptor must be written with 32 bits words. If the CAAM addressing mode is 64 bits, all addresses in the descriptor must be coded with 64 bits (hence size of the descriptor is bigger in this case than in the case of CAAM addressing mode of 32 bits).

cneveux commented on core/drivers/crypto/caam/caam_desc.c:
```diff
@@ -0,0 +1,90 @@
+void desc_add_word(uint32_t *desc, uint32_t word)
+{
+	uint32_t len = GET_JD_DESCLEN(caam_read_val((void *)desc));
+	uint32_t *last = desc + len;
```
len is the number of entries but it's define as length (length of the descriptor) in the CAAM HW specification.

Emantor commented on core/arch/arm/plat-imx/crypto_conf.mk:
```diff
@@ -0,0 +1,69 @@
+# DTB Modification
+#
+ifeq ($(CFG_DT), y)
+CFG_CAAM_DT ?= y
```
Who Shadow `CFG_DT` with `CFG_CAAM_DT`? All occurences of `CFG_CAAM_DT` can probably be exchanged for `CFG_DT`.

cneveux replied:
Because we want to keep the granularity in the CAAM Driver to use or not the Device Tree. Some customers are not using the Device Tree as provided by default in Linux.

Emantor commented on core/arch/arm/plat-imx/crypto_conf.mk:
```diff
@@ -0,0 +1,69 @@
+$(call force, CFG_JR_BLOCK_SIZE,0x1000)
+
+$(call force, CFG_JR_INDEX,0)  # Default JR index used
+$(call force, CFG_JR_IRQ,137)  # Default JR IT Number (105 + 32) = 137
```
These can probably be removed and carried as regular defines until support for another platform is added.

cneveux replied:
It's not possible because we already support multi-platforms (like LS)

Emantor commented on core/arch/arm/plat-imx/crypto_conf.mk:
```diff
@@ -0,0 +1,69 @@
+$(call force, CFG_WITH_SOFTWARE_PRNG,n)
+else
+$(call force, CFG_WITH_SOFTWARE_PRNG,y)
+endif
```
This can be simplified to `$(call force, CFG_WITH_SOFTWARE_PRNG,n)`.

cneveux replied:
I'll prefer to keep the `#ifeq ($CFG_CRYPTO_RNG_HW),y)` check to force the `CFG_WITH_SOFTWARE_PRNG` to `n`.
And yes, you're right I'll remoce the `$(call force, CFG_WITH_SOFTWARE_PRNG,y)`

Emantor commented on core/drivers/crypto/caam/hal/ls/hal_clk.c:
```diff
@@ -0,0 +1,19 @@
+void hal_clk_enable(bool enable __unused)
+{
+}
+
```
All the `hal/ls/*` files can probably be dropped since this PR only adds support for imx. They can be readded later with `ls` support.

cneveux replied:
It can not be removed as it's a platform already supported and we have aligned with the team to ensure that we are upstreaming support to enable them too.

cneveux commented:
> for commit "core: io: Add 32/64 little endian function":
> `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
> 
> Nitpicking: suggest to rephrase commit message (in the model of [685204e](https://github.com/OP-TEE/optee_os/commit/685204eb3f2f4162fd7e4ec76e7089cc25de5246)):
> 
> ```
> core: io: add {get,put}_le{32,64}()
> 
> Add 32 and 64 bit little endian put/get functions
> 
> Signed-off-by: Cedric Neveux <cedric.neveux@nxp.com>
> ```

ok

cneveux commented on core/crypto/crypto.c:
```diff
@@ -14,30 +14,39 @@
+	 */
+	res = drvcrypt_hash_alloc_ctx(&c, algo);
+
+	if (res) {
```
not sure that's what was suggested. @jenswi-linaro asked to have inline function returning TEE_ERROR_NOT_IMPLEMENTED if the CFG_CRYPTO_DRIVER is not defined. But as you want I'll do the change.

etienne-lms replied:
If `drvcrypt_hash_alloc_ctx()` fails with `TEE_ERROR_OUT_OF_MEMORY`, there is obviously not point in trying to allocate another context. The same applies to many error codes. I think Jens suggested to stub `drvcrypt_hash_alloc_ctx()` with returning `TEE_ERROR_NOT_IMPLEMENTED` for that very purpose.

cneveux commented on core/arch/arm/plat-imx/conf.mk:
```diff
@@ -285,3 +286,21 @@ CFG_SHMEM_SIZE ?= 0x00200000
+# If NXP CAAM Driver is supported, the Crypto Driver interfacing
+# it with generic crypto API can be enabled.
+CFG_CRYPTO_DRIVER ?= y
+# Crypto Driver Debug
```
I prefer to keep it as information

cneveux commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,125 @@
+	ret = caam_jr_complete();
+
+	if (ret) {
+		EMSG("CAAM initialization failed\n");
```
ok

cneveux commented:
[code formatting] Run Clang-format on the following commits:
core: driver: generic resources for crypto device driver
core: driver: implementation NXP CAAM Driver

jforissier commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,130 @@
+ * @brief   CAAM Global Controller.\n
+ */
+
+/* Global includes */
```
Agreed and I also do not like the separation `/* Global includes */` vs. `/* Local includes */`. We don't normally do that, and if we accept this here, this will undoubtedly be used as a template by other contributors.

cneveux replied:
ok, I'll remove this split that was helpful for me.

jforissier commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,120 @@
+ * Returns:
+ * TEE_SUCCESS              Success
+ * TEE_ERROR_GENERIC        Generic Error (driver init failure)
+ * TEE_ERROR_NOT_SUPPORTED  Driver not supported
```
Am I the only one to think this is bad? (documenting the return codes like this)

jenswi-linaro replied:
You mean this level of details which easily could become outdated and thus do more harm than good?

etienne-lms replied:
I think it is useless if not simply wrong.

cneveux replied:
What is usual practice?

etienne-lms replied:
When using already defined value, prefer centralized their description at one place. This applies to `TEE_ERROR_*` (ref is the GPD TEE spec) as well as your `CAAM_*` status values (one should rather refer to `caam_status.h` for the description).

Here, as returning a `TEE_Result` typed value, the code implicitly expect standard `TEE_Result` values are used, hence I would suggest to simply not describe the return value:
```
/* Crypto driver initialization */
static TEE_Result crypto_driver_init(void)
{
	(...)
}

/* Crypto driver late initialization to complete on-going CAAM operations */
static TEE_Result init_caam_late(void)
{
	(...)
}
```

cneveux replied:
Fine I'll review all functions

etienne-lms commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,759 @@
+ *
+ * Returns:
+ * 0    CAAM is no more busy
+ * (-1) CAAM is still busy
```
suggest to use `CAAM_NO_ERROR` and `CAAM_BUSY`. Same for other functions above that could use `CAAM_*` status values.

cneveux replied:
ok, thanks

etienne-lms commented on core/drivers/crypto/caam/caam_ctrl.c:
```diff
@@ -0,0 +1,120 @@
+{
+	int ret = 0;
+
+	ret = caam_jr_complete();
```
Why completing JR jobs during late init? Does this implicitly expect no other "late init" function is allowed to run JR jobs?

cneveux replied:
This is to ensure that there is no pending jobs running in CAAM HW before leaving OPTEE boot. This is mandatory on certain platforms to continue the authentication boot processus.

etienne-lms replied:
It's not clear when new jobs are allowed to be started. `caam_ctrl.c` may be used by other platforms not following your boot process, no?
(note that while i push this comment, i know nothing about such platforms. Feel free to discard my comment if you think the architecture/implementation is safe)

cneveux replied:
@etienne-lms 
Well, I would not explain how to CAAM HW is working but CAAM is like a coprocessor that is running jobs while there is one present in the input queue. When a job is complete CAAM pushes information in the output queue and we have to take status completion from this queue.
As the CAAM is running independently and because of the different boot process on NXP devices, we have to ensure that there is no running job in the CAAM to continue.

caam_ctrl.c is specific to the CAAM HW and the driver itself. I don't see the interest for another platform to re-use it if it's HW doesn't work like ours.

I really appreciate the review you are doing, that help me to fix hidden issues. For sure, the architecture/implementation is safe in the sense where it's working since years on different OPTEE version and other internal libraries. We can still have bug because of the integration of the driver in OPTEE but in terms of architecture I think we are clean now.

cneveux commented:
@jenswi-linaro, @etienne-lms, @jforissier 

Can I ask you if there is more comments on this PR? Do you approved it?

Thanks and Regards,

Cedric

jenswi-linaro commented:
@cneveux, I'm not done yet I was just taking a break since @etienne-lms was bombarding you with comments. I'll resume again.

By the way, often there's no notification when a PR is just force pushed, a comment is needed to to make sure that it's noticed.

cneveux commented:
> @cneveux, I'm not done yet I was just taking a break since @etienne-lms was bombarding you with comments. I'll resume again.
> 
> By the way, often there's no notification when a PR is just force pushed, a comment is needed to to make sure that it's noticed.

@jenswi-linaro 
Ok, thanks. As I'll be in vacation from Aug, 15th to Aug, 26th, I'll certainly not have time to review and submit a new patch before leaving.

jenswi-linaro commented:
@cneveux, thanks for letting me know. I hope you'll have a good vacation. I'll make sure there's some comments waiting for you when you're back. :-)

jenswi-linaro commented on core/drivers/crypto/crypto_api/drvcrypt_init.c:
```diff
@@ -0,0 +1,59 @@
+#include <drvcrypt.h>
+#include <initcall.h>
+#include <trace.h>
+
```
I'm not sure the interface implemented in this file is needed. It only seems safe to call during early boot, before exiting to normal world the first time. Type safety is lost. The caller of `drvcrypt_getmod()` must know the type of the point it gets. All this could be replaced by:
``` C
const struct drvcrypt_hash *drvcrypt_hash_ops;
const struct drvcrypt_mac *drvcrypt_mac_ops;
const struct drvcrypt_cipher *drvcrypt_cipher_ops;
...
```
with retained type safety. Perhaps I'm missing something.

If you think it's important to keep this, do so. Once all the PRs are merged we can evaluate this again.

cneveux replied:
This interface is to register to HW Driver module in the crypt_algo array. The registration is done during the HW Driver initialization function of the HW capabilities. To be generic and to prevent optional compilation per device type, we read HW registers to know if CAAM support or not the algorithm to register. If not, we keep NULL in the crypt_algo and so the HW driver is not called.
For sure you can not see to need with the HASH but for other operation it will be usefull

jenswi-linaro commented on core/drivers/crypto/crypto_api/drvcrypt_init.c:
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2018-2019 NXP
+ *
+ * Brief   This driver interfaces TEE Internal API by implementing
```
"TEE Internal API" is an unfortunate wording as that's something different for us who have worked a bit with the GP APIs.

cneveux replied:
ok I'll reword it.

jenswi-linaro commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,231 @@
+ */
+struct crypto_hash {
+	struct crypto_hash_ctx hash_ctx; /* Crypto Hash API context */
+
```
Empty line not needed.

jenswi-linaro commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,231 @@
+ * Format the HASH context to keep the reference to the
+ * operation driver
+ */
+struct crypto_hash {
```
Is this wrapper layer really needed? Is anything missing with `struct crypto_hash_ctx` which is stopping `drivers/crypto/caam/hash/caam_hash.c` from implementing a `struct crypto_hash_ctx` instead? Am I missing something?

cneveux replied:
For sure, now it's no more needed as we reworked everything and removed the original checks. I can remove it but for the next crypto driver, I'll have the same structure and to be compliant I would prefer to keep this method. Now if you think that it will add complexity and overhead I can rework it to remove the `struct drvcrypt_hash`.
Please let me know.

jenswi-linaro replied:
What do you mean by the next crypto driver? Compliant with what, in which context?

If it's only here for historical reasons I'd rather see it removed since it adds another level of indirection and makes it harder to follow the code.

cneveux replied:
I mean the implementation for the Symmetric, Asymmetric and key derivation using the HW driver. There is the same wrapping operation structure used to reformat the crypto_xxx parameters in a data structure.
Anyway, for the Hash I'll remove it and we will see after for the others.

cneveux replied:
I added a new commit (last one in the list) to the PR implementing this change. I'll squashed files after review and acceptance

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,768 @@
+{
+	uint8_t idx = 0;
+
+	JR_TRACE("Job cancel 0x%"PRIx32"", jobId);
```
@etienne-lms, when I saw this with the lack of spaces this first time I thought the same thing. However, after seeing this for a while it is actually a bit easier to read. I guess it's because it's slightly more compact and also looks a bit more like a format directive.

I don't mind if we keep skipping the spaces around `PRI*` macros in strings, I've started to use this style myself.

jenswi-linaro commented on core/drivers/crypto/caam/caam_jr.c:
```diff
@@ -0,0 +1,709 @@
+static uint32_t do_jr_dequeue(uint32_t waitJobIds)
+{
+	uint32_t retJobId = 0;
+
```
Empty line not needed.

cneveux replied:
ok

jenswi-linaro commented on core/drivers/crypto/caam/include/caam_hal_clk.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef __CAAM_HAL_CLK_H__
+#define __CAAM_HAL_CLK_H__
+
+#include <stdlib.h>
```
Including `<types_ext.h>` would be a better choice.

jenswi-linaro commented on core/drivers/crypto/caam/include/caam_hal_ctrl.h:
```diff
@@ -0,0 +1,34 @@
+#ifndef __CAAM_HAL_CTRL_H__
+#define __CAAM_HAL_CTRL_H__
+
+#include <utee_defines.h>
```
Including <types_ext.h> would be a better choice.

jenswi-linaro commented on core/drivers/crypto/caam/utils/utils_delay.c:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @time  Time in microsecond
+ */
+void caam_udelay(uint32_t time)
```
How about using `udelay()` instead?

Emantor replied:
`udelay()` currently uses the core internal timer, which is not available on Cortex-A9 cores, i.e. i.MX6Q.

jenswi-linaro replied:
Aha, missed that. Perhaps the platform should be able to override `udelay()`. Because this implementation doesn't seem very suitable for generic code.

cneveux replied:
As this is specific to NXP CAAM driver I prefer to not change it and it will be reviewed later when we will finish the NXP i.MX implementation support for all device type (MX6/7/8).

jenswi-linaro replied:
OK

jenswi-linaro commented:
There's quite a few cases of camel case in especially the commit "core: driver: implementation NXP CAAM Driver". `enum CAAM_Status` is acceptable, at least to me, but the rest should preferably be fixed.

jenswi-linaro commented:
There's quite a few new checkpatch warnings now that checkpatch is run with `--strict`.

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,231 @@
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
+	if (hash->op) {
```
The free operation in the driver is not mandatory. Hence the check of the pointer is needed to call the driver function.

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,231 @@
+	struct crypto_hash *hash_src = to_hash_ctx(src_ctx);
+	struct crypto_hash *hash_dst = to_hash_ctx(dst_ctx);
+
+	if (hash_src->op)
```
my mistake I missed it

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,231 @@
+static TEE_Result hash_init(struct crypto_hash_ctx *ctx)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+
```
ok

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,231 @@
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
+	if ((!data) && (len != 0))
```
ok, I'll fix it

cneveux commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,231 @@
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
+	if ((!digest) || (!len))
```
ok

cneveux commented on core/drivers/crypto/crypto_api/include/drvcrypt_hash.h:
```diff
@@ -0,0 +1,31 @@
+	TEE_Result (*update)(void *ctx, const uint8_t *data, size_t len);
+	/* Finalize the hashing operation */
+	TEE_Result (*final)(void *ctx, uint8_t *digest, size_t len);
+
```
ok

cneveux commented on core/drivers/crypto/caam/caam_desc.c:
```diff
@@ -0,0 +1,82 @@
+/* Return lower 32 bits of physical address */
+#define PHYS_ADDR_LO(phys_addr) (uint32_t)(((uint64_t)phys_addr) & UINT32_MAX)
+
+uint32_t desc_get_len(uint32_t *desc)
```
I see your point. I'll add the prefix

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,218 @@
+{
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
```
can remove this comment

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,218 @@
+ * Copy Software Hashing Context
+ *
+ * @src_ctx  Reference the API context source
+ * @dst_ctx  [out] Reference the API context destination
```
swap lines (effective args ordering)

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,218 @@
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	/* Check the parameters */
```
remove comment
same for all `/* Check the parameters */` in this file.

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,218 @@
+ * Finalize the Hash operation
+ *
+ * @ctx     Reference the API context
+ * @len     Digest buffer length
```
swap the 2 last lines.

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,218 @@
+ * @algo  Algorithm
+ * @id    [out] Hash Algorithm internal ID
+ */
+static struct drvcrypt_hash *do_check_algo(uint32_t algo, uint8_t *id)
```
suggest to move this below line 178, right above `drvcrypt_hash_alloc_ctx()`.

etienne-lms commented on core/drivers/crypto/crypto_api/include/drvcrypt.h:
```diff
@@ -0,0 +1,57 @@
+ * Cryptographic module registration
+ *
+ * @algo_id  Crypto index in the array
+ * @ops      Reference to the cryptographic module
```
suggest: "Registered operation table of the module"

etienne-lms commented on core/drivers/crypto/crypto_api/include/drvcrypt.h:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Cryptographic module registration
+ *
+ * @algo_id  Crypto index in the array
```
could suggest a more agnostic way (I.e. "ID of the registered crypto algorithm")

etienne-lms commented on core/drivers/crypto/crypto_api/include/drvcrypt_hash.h:
```diff
@@ -0,0 +1,30 @@
+ * Crypto Library Hash driver operations
+ */
+struct drvcrypt_hash {
+	/* Allocates of the Software context */
```
s/of the/the/

Some suggested rephrasing:
"Allocate operation context"
"Free operation context"
"Initialize operation context"
"Update  operation context"
"Finalize  operation context"
"Copy operation context"
...or maybe even no inline comment.

etienne-lms commented on core/include/crypto/crypto_impl.h:
```diff
@@ -27,6 +27,18 @@ struct crypto_hash_ops {
 			   struct crypto_hash_ctx *src_ctx);
 };
 
+#ifdef CFG_CRYPTO_DRIVER
```
Should be `CFG_CRYPTO_DRV_HASH`?

etienne-lms commented on core/drivers/crypto/crypto_api/include/drvcrypt.h:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Cryptographic buffer type
+ */
+struct drvcrypt_buf {
```
unused. remove.

etienne-lms commented on core/drivers/crypto/crypto_api/hash/hash.c:
```diff
@@ -0,0 +1,218 @@
+	if (hash->op && hash->op->alloc_ctx)
+		ret = hash->op->alloc_ctx(&hash->ctx, hash_id);
+
+	if (ret != TEE_SUCCESS) {
```
If module has no `alloc_ctx` handler, this will fail with `TEE_ERROR_NOT_IMPLEMENTED`. 
Could default init `ret = TEE_SUCCESS;`, or change a bit the code.

etienne-lms commented:
@cneveux, I suggest you create a P-R for the preliminary changes (all already reviewed) so that they can be merged:
* core: arm64 Add sev/wfe inline function  …
* plat-imx: Add CRM registers mapping  …
* core: add DT function to read/modify DT  …
* core: kernel: move DT pack at the end of boot  …
* core: io: add {get/put}_le{32/64}()  …
 * core: arm64: add read_ctr_el0 function  …

Also minor comment regarding commit "core: driver: generic resources for crypto device driver ":
(`core/drivers/crypto/crypto_api/`)`drvcrypt_init.c` could be renamed `drvcrypt.c`.

jenswi-linaro commented:
> @cneveux, I suggest you create a P-R for the preliminary changes (all already reviewed) so that they can be merged...

And for the commits left perhaps also a new PR because it's quite slow to browse this PR now.

cneveux commented:
@etienne-lms, @jenswi-linaro,

As suggested, created a PR #3222 with:

- core: arm64 Add sev/wfe inline function …
- plat-imx: Add CRM registers mapping …
- core: add DT function to read/modify DT …
- core: kernel: move DT pack at the end of boot …
- core: io: add {get/put}_le{32/64}() …
- core: arm64: add read_ctr_el0 function …

etienne-lms commented:
Thanks @cneveux.
When updating the drvcrypt serie, please straight squash 
"core: driver: Remove drvcrypt_hash ops"
and fixup "core: driver: generic resources for crypto device driver"
 as you [suggested](https://github.com/OP-TEE/optee_os/pull/2962#discussion_r317983367).
